Sources/PT.PM.Common/Reflection/ReflectionCache.cs

```diff
@@ -1,53 +1,173 @@
 ﻿using PT.PM.Common.Nodes;
 using System;
-using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
 using System.Reflection;
+using System.Runtime.Serialization;
+using MessagePack;
 using Newtonsoft.Json;
 
 namespace PT.PM.Common.Reflection
 {
     public static class ReflectionCache
     {
-        private static ConcurrentDictionary<Type, PropertyInfo[]> ustNodeProperties
-            = new ConcurrentDictionary<Type, PropertyInfo[]>();
+        private static readonly HashSet<Type> registeredTypes;
+        private static readonly Dictionary<string, Type> kindClassType;
+        private static readonly Dictionary<Type, (byte Type, PropertyInfo[] Properties)> typeSerializableProperties;
+        private static readonly Type[] ustTypes;
 
-        private static Lazy<Dictionary<string, Type>> UstKindClassType =
-            new Lazy<Dictionary<string, Type>>(() =>
+        static ReflectionCache()
+        {
+            registeredTypes = new HashSet<Type>();
+            kindClassType = new Dictionary<string, Type>();
+            typeSerializableProperties = new Dictionary<Type, (byte NodeType, PropertyInfo[] Properties)>();
+
+            Assembly ustAssembly = typeof(Ust).Assembly;
+            var types = (UstType[]) Enum.GetValues(typeof(UstType));
+            ustTypes = new Type[types.Length];
+            var assemblyTypes = ustAssembly.GetTypes().Where(type => type.IsSubclassOf(typeof(Ust))).ToList();
+
+            foreach (UstType type in types)
             {
-                var result = new Dictionary<string, Type>();
-                foreach (Assembly assembly in AppDomain.CurrentDomain.GetAssemblies())
+                string typeStr = type.ToString();
+                Type ustType = null;
+                foreach (Type assemblyType in assemblyTypes)
                 {
-                    if (assembly.IsActual())
+                    if (assemblyType.Name.Equals(typeStr))
                     {
-                        foreach (Type type in assembly.GetTypes()
-                            .Where(myType => myType.IsClass && !myType.IsAbstract &&
-                            myType.GetInterfaces().Contains(typeof(IUst))))
-                        {
-                            result.Add(type.Name.ToLowerInvariant(), type);
-                        }
+                        ustType = assemblyType;
```

Comment:
This will overwrite `ustType` with the last `assemblyType` satisfying the condition if there are several of those. Is it correct behaviour? Why? Or should there be an early exit?

Reply:
Thanks, we should add an early exit statement. Also, I think it makes sense to add more unit tests for reflection checking.

Reply:
Use LINQ! It makes code harder to get wrong! :smile: